### PR TITLE
Nvtx transform

### LINKED ISSUE #2712: NVTX in MONAI
**Is your feature request related to a problem? Please describe.**
Nsight Systems provide profiling of a deep learning workflow. Apart from its predefined tracking components, a user can employ NVIDIA Tool Extension (NVTX) to mark event and profile specific functions. For MONAI transforms, it is not possible to add these markers directly since it is using a compose pipeline.

**Describe the solution you'd like**
Wrapping NVTX functions as MONAI transformations enables profiling of any arbitrary MONAI transformation by simply adding these transformation into the MONAI pipeline.

### PR DESCRIPTION
**Pull request by drbeh**
Fixes #2712 

### Description
This PR implements transformation for NVIDIA Tool Extension (NVTX) functions.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by drbeh** [comment]
Hi @wyli and @Nic-Ma, 
These transformations only have effect when they are run with `nsys profile`, and also the results need to be visualized by "Nsight Systems" in order to be checked, so I am not sure how to write a test case to work with MONAI CI/CD. I have tested them in my local machine, and they are working as expected. What do suggest to do? Is local testing fine?

**Comment by drbeh** [comment]
@wyli and @Nic-Ma, I added some unittests to make sure that these tranforms does not alter anything when they are used alone or in conjunction with other array or dictionary transforms.

**Review by Nic-Ma — COMMENTED** [review]
Hi @drbeh ,

Thanks for bringing this interesting feature to MONAI.
To verify and test the APIs, I would suggest you to add a simple tutorial notebook in the MONAI tutorials repo, and put the link in the doc-string of this module. Then users can easily get start to use it and we can also easily verify & test it.

Thanks.

**Comment by Nic-Ma** [comment]
@ericspod @rijobro @wyli ,

Do you guys have any other comments?

Thanks in advance.

File: monai/transforms/nvtx.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    def __init__(self, msg: str) -> None:
+        self.msg = msg
+        self.depth = None
+
+    def __call__(self, data):
+        self.depth = _nvtx.rangePushA(self.msg)
+        return data
+
+
+class RandRangePush(RandomizableTransform):
```
[the remark this answers is not in the document]
Comment (drbeh) [reply monai/transforms/nvtx.py:41]: You're observation is definitely right! The reason for these randomizable versions is to not alter any transform that they are enclosing. For instance, in `SmartCacheDataset` and `CachedDataset`, only the initial non-randomizable transforms are being cached, so their corresponding NVTX transforms should be correctly cached or not cached.

File: monai/transforms/nvtx.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    Pops a range off of a stack of nested range spans.
+    Stores zero-based depth of the range that is ended.
+    """
+
+    def __call__(self, data):
+        _nvtx.rangePop()
+        return data
+
+
+class RandRangePop(RandomizableTransform):
```
[the remark this answers is not in the document]
Comment (drbeh) [reply monai/transforms/nvtx.py:71]: See the comment on `RandRangePush`. Thanks

File: monai/transforms/nvtx.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    def __init__(self, msg: str) -> None:
+        self.msg = msg
+
+    def __call__(self, data):
+        _nvtx.markA(self.msg)
+        return data
+
+
+class RandMark(RandomizableTransform):
```
[the remark this answers is not in the document]
Comment (drbeh) [reply monai/transforms/nvtx.py:101]: See the comment on `RandRangePush`. Thanks

File: tests/test_nvtx_transform.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    np.random.randn(3, 3),
+]
+TEST_CASE_1 = [
+    {"image": np.random.randn(3, 3)},
+]
+
+
+class TestNVTXTransforms(unittest.TestCase):
+    @parameterized.expand([TEST_CASE_0, TEST_CASE_1])
+    @unittest.skipUnless(has_nvtx, "CUDA is required for NVTX!")
```
[the remark this answers is not in the document]
Comment (drbeh) [reply tests/test_nvtx_transform.py:47]: It doesn't have a new dependency. It depends on PyTorch that is being compiled with CUDA. So these tests should be skipped in all non-CUDA environments, and should be executed on our CUDA environments. Am I missing anything?

File: tests/test_nvtx_transform.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                RangePop(),
+                Mark("Mark: Transform Ends!"),
+            ]
+        )
+        output = transforms(input)
+        self.assertIsInstance(output, torch.Tensor)
+        try:
+            np.testing.assert_array_equal(input, output)
+        except AssertionError:
+            np.testing.assert_array_equal(input, Flip()(output.numpy()))
```
[the remark this answers is not in the document]
Comment (drbeh) [reply tests/test_nvtx_transform.py:82]: Sure, I'll update it tomorrow (Eastern US). ;)

File: monai/transforms/nvtx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/nvtx.py:41]: OK, I think maybe this class can be subclass of non-random version, no need to write duplicated code for it?
```py
class RandRangePush(RangePush, RandomizableTransform)
```

File: tests/test_nvtx_transform.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply tests/test_nvtx_transform.py:47]: OK, I see. Glad to know that.
Thanks.

**Comment by drbeh** [comment]
> @ericspod @rijobro @wyli ,
> 
> Do you guys have any other comments?
> 
> Thanks in advance.

I'd appreciate it if you can give me any comment that you have.
Thanks.

**Comment by wyli** [comment]
Thanks for the PR, my feeling is that the NVTX tools are not well-known in our community. Shall we have some tutorials first, then revisit this PR and move the reusable code blocks from the tutorials into the core codebase? I don't know why these are wrapped as transforms instead of network layers, but I guess tutorials would clarify this easily.

**Comment by drbeh** [comment]
> Thanks for the PR, my feeling is that the NVTX tools are not well-known in our community. Shall we have some tutorials first, then revisit this PR and move the reusable code blocks from the tutorials into the core codebase? I don't know why these are wrapped as transforms instead of network layers, but I guess tutorials would clarify this easily.

@wyli, tutorial would be great to teach the community on how to use Nsight System and NVTX. 
Nsight is able to profile the network by default, so there is no need to wrap them as network. To profile a chain of transforms, you should either put NVTX in MONAI source code for each transform, or wrap them as transforms to be able to put them where is need in a transform pipeline. This PR is doing the latter, which is not intrusive and does not affect any other components in MONAI.

**Comment by Nic-Ma** [comment]
> Hi @drbeh ,
> 
> Thanks for bringing this interesting feature to MONAI.
> To verify and test the APIs, I would suggest you to add a simple tutorial notebook in the MONAI tutorials repo, and put the link in the doc-string of this module. Then users can easily get start to use it and we can also easily verify & test it.
> 
> Thanks.

Hi @drbeh ,

As I said in previous comment, I agree with @wyli , we should make a tutorial first and put the link in the doc-string of this module.

Thanks.

**Comment by drbeh** [comment]
> > Hi @drbeh ,
> > Thanks for bringing this interesting feature to MONAI.
> > To verify and test the APIs, I would suggest you to add a simple tutorial notebook in the MONAI tutorials repo, and put the link in the doc-string of this module. Then users can easily get start to use it and we can also easily verify & test it.
> > Thanks.
> 
> Hi @drbeh ,
> 
> As I said in previous comment, I agree with @wyli , we should make a tutorial first and put the link in the doc-string of this module.
> 
> Thanks.

@wyli, @Nic-Ma, I definitely agree with having a tutorial on this but the tutorial will depend on these components, so shouldn't we first merge this and then use this in a tutorial?!

Similar to the other components where we first have implemented in MONAI (like other transforms and models), and then created tutorial for spleen segmentation for example, right?

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

I would suggest you to develop a tutorial notebook in the `tutorials` repo, then let's merge these 2 PRs together.

Thanks.

**Comment by drbeh** [comment]
> Hi @drbeh ,
> 
> I would suggest you to develop a tutorial notebook in the `tutorials` repo, then let's merge these 2 PRs together.
> 
> Thanks.

Sure, I can do that.

**Comment by drbeh** [comment]
Hi @Nic-Ma , @wyli

I have created a notebook and necessary codes for profiling digital pathology transformation pipeline using these NVTX transforms, and NVIDIA Nsight System. Please check it out: https://github.com/Project-MONAI/tutorials/pull/305 

Thanks

**Review by Nic-Ma — APPROVED** [review]
Thanks for the tutorial, looks good to me now.

File: monai/transforms/nvtx.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    """
+    Pushes a range onto a stack of nested range span (RandomizableTransform).
+    Stores zero-based depth of the range that is started.
+
+    Args:
+        msg: ASCII message to associate with range
+    """
+
+
+class RangePop(Transform):
```
Comment (wyli) [inline monai/transforms/nvtx.py:76]: according to the tutorial, I think those push/pop operations should be implemented as a python context manager for better readability, what do you think?

Comment (drbeh) [reply monai/transforms/nvtx.py:76]: @wyli, you're right but the reason that we have separate push/pop transforms is that the range that they defines is not limited to a transform and can include multiple transforms.

Comment (drbeh) [reply monai/transforms/nvtx.py:76]: I'll implement the context manager too.  #2754

**Pull request merged**